src/AtomsGaffer/AtomsVariationReader.cpp

```diff
@@ -744,22 +745,43 @@ public :
     void loadAgentTypeMesh( Atoms::AgentTypeVariationCPtr agentTypePtr, const std::string& agentTypeName )
     {
         auto geoNames = agentTypePtr->getGeometryNames();
-        for ( size_t geoId = 0; geoId != geoNames.size(); ++geoId )
-        {
-            auto geoPtr = agentTypePtr->getGeometryPtr( geoNames[geoId] );
-            if ( !geoPtr )
-                continue;
+        size_t nbGeos = geoNames.size();
 
-            if ( geoPtr->getGeometryFile().find(".groom") != std::string::npos )
-                continue;
+        std::vector<AtomsPtr<AtomsCore::MapMetadata>> atomsMeshes(nbGeos);
+        std::vector<Atoms::VariationGeometryCPtr> atomsGeoPtrs(nbGeos, nullptr);
 
-            auto inGeoMap = Atoms::loadMesh( geoPtr->getGeometryFile(), geoPtr->getGeometryFilter() );
-            if ( !inGeoMap )
+        // Load atoms meshes in parallel
+        tbb::blocked_range<size_t> tbb_range(0, nbGeos, 1);
+        tbb::parallel_for(tbb_range, [&](const tbb::blocked_range<size_t>& range) {
```

Comment:
You have to be careful with tbb parallelization inside gaffer nodes because of the way the compute cache policies work. Here's an example:

https://github.com/GafferHQ/gaffer/blob/15f6530800a99028fd60dbfc3cc1541d0bc139c8/src/GafferScene/PrimitiveSampler.cpp#L371

Reply:
There are two different considerations here :

1. TBB exception handling. Unless told otherwise, TBB will silently cancel your `parallel_for` if an exception is thrown in another task inside the same task tree - which as far as you are concerned could be completely unrelated to what you are doing. This is what you need the `task_group_context` that Andrew linked for - it tells TBB that your `parallel_for` is a distinct operation that shouldn't be affected by exceptions thrown in any others.
2. Deadlock due to TBB task stealing. TBB loves to wander off and do whatever TBB tasks it can squeeze in, so the wait for your `parallel_for` can cause the execution of completely arbitrary TBB tasks elsewhere in Gaffer. Gaffer performs does some operations behind mutexes, so if your `parallel_for` is behind that mutex, and the arbitrary task TBB chooses to do needs that mutex, you get deadlock. _This_ is what the cache policies Andrew mentioned are for. By declaring a cache policy of either `TaskIsolation` or `TaskCollaboration`, you avoid the deadlock by making sure TBB won't do any outer tasks unrelated to your `parallel_for`. `TaskCollaboration` has more overhead but is smart enough to allow multiple waiting threads to collaborate on the construction of the engine if they all pull on the engine plug concurrently. There's an example of this in [`FilterResults`](https://github.com/GafferHQ/gaffer/blob/42474683e7e35d090ecf55ca026cfb3f70cdafb9/src/GafferScene/FilterResults.cpp#L166-L169).

Yes, there is an unfortunate overloading of the term "isolation" here. This comes from TBB, where `task_group_context::isolated` means "not cancelled by exceptions elsewhere" and `this_task_arena::isolate` means "won't wander off and perform unrelated tasks".

Reply:
I was not aware of that. Thank you for the clear explanation. :) Just to be sure, that means that setting the cache policy on the engine data output to TaskCollaboration will allow the currently running threads (the ones from which the engineDataPlug compute method are called from)  to wait and pick up the subtasks (atoms mesh loading) while they are waiting for the output to be computed? Does that mean that setting it to TaskIsolation will prevent the subtasks to have access to all cores, or that the subtasks will simply need to create more threads instead of reusing the ones waiting, which causes more overhead?

Reply:
> Just to be sure, that means that setting the cache policy on the engine data output to TaskCollaboration will allow the currently running threads (the ones from which the engineDataPlug compute method are called from) to wait and pick up the subtasks (atoms mesh loading) while they are waiting for the output to be computed?

Yes.

> Does that mean that setting it to TaskIsolation will prevent the subtasks to have access to all cores?

Yes, if other tasks/threads also need the result from `engineDataPlug()`, they would just wait uselessly for it in the TaskIsolation case. TaskIsolation solves the deadlock problem, but does nothing to aid performance. In an ideal world, TaskCollaboration would be the only option, but it has additional overhead so we are keeping TaskIsolation around as an option until we're happy that collaboration is suited for everything - more recent TBBs seem to have a new mechanism that might help us implement it more efficiently.

In your case, I would expect the engine loading to be heavy enough and infrequent enough that TaskCollaboration would be the way to go.

Reply:
Ok, thanks!__